research/cifar10/model.py

```diff
@@ -45,3 +48,128 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
         x = self.fc2(x)
 
         return x
+
+
+class ConvNetFeatureExtractor(Module):
+    def __init__(
+        self,
+        in_channels: int,
+        use_bn: bool = True,
+    ) -> None:
+        super().__init__()
+
+        self.conv1 = Conv2d(in_channels, 32, 5, padding=2)
+        self.conv2 = Conv2d(32, 64, 5, padding=2)
+        self.use_bn = use_bn
+        if use_bn:
+            self.bn1 = BatchNorm2d(32)
+            self.bn2 = BatchNorm2d(64)
+
+        self.relu = ReLU(inplace=True)
+        self.maxpool = MaxPool2d(2)
+        self.flatten = Flatten()
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+
+        x = self.bn1(self.conv1(x)) if self.use_bn else self.conv1(x)
+        x = self.maxpool(self.relu(x))
+        x = self.bn2(self.conv2(x)) if self.use_bn else self.conv2(x)
+        x = self.maxpool(self.relu(x))
+        x = self.flatten(x)
+
+        return x
+
+
+class ConvNetClassifier(Module):
+    def __init__(
+        self,
+        h: int = 32,
+        w: int = 32,
+        hidden: int = 2048,
+        class_num: int = 10,
+        dropout: float = 0.0,
+    ) -> None:
+        super().__init__()
+
+        self.fc1 = Linear((h // 2 // 2) * (w // 2 // 2) * 64, hidden)
+        self.fc2 = Linear(hidden, class_num)
+
+        self.relu = ReLU(inplace=True)
+        self.dropout_layer = nn.Dropout(p=dropout)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = self.dropout_layer(x)
+        x = self.relu(self.fc1(x))
+        x = self.dropout_layer(x)
+        x = self.fc2(x)
+
+        return x
+
+
+class ConvNetFendaClassifier(ParallelSplitHeadModule):
+    def __init__(
+        self,
+        join_mode: ParallelFeatureJoinMode,
+        h: int = 32,
+        w: int = 32,
+        hidden: int = 2048,
+        class_num: int = 10,
+        dropout: float = 0.0,
+    ) -> None:
+        super().__init__(join_mode)
+
+        # Times 2 because we'll be concatenating the inputs of two feature extractors
+        self.fc1 = Linear((h // 2 // 2) * (w // 2 // 2) * 64 * 2, hidden)
```

Comment:
Nice! I didn't know about the `//` operator.

Reply:
Credit to @sanaAyrml for this usage.